drivers/ethernet/phy_xlnx_gem.h

```diff
@@ -0,0 +1,119 @@
+/*
```

Comment:
The #34718 is pending and is providing MDIO and PHY driver support. Is it possible to utilize that generic support here?

Reply:
I'll have a look into that tomorrow. At first glance it seemed to me that up until now, all the drivers supporting PHY management all did their own thing. If that interface provides link state and link speed information and is suitable for interrupt-driven operation where the interrupt line is actually that of the controller itself plus a corresponding bit set in the interrupt status register, plus it allows the integration of PHY types not previously supported, I guess that might work.

Reply:
I've had a first look at this PR and at the time being, it doesn't look suitable:
- It's unclear how interrupt handling is going to work. If this is supposed to be implemented by using a dedicated ISR attached to an interrupt line specified by an interrupts property within the PHY's device tree entry, as mentioned above, that's incompatible with the way the GEM signals the PHY interrupt.
- Gigabit is not considered.
- The API doesn't seem to have any method of specifying what shall actually be advertised to the link partner. The current generic MII driver implementation triggers auto-negotiation upon the link going up, looks at the link partner capability register, and returns the best possible link from that set of data. I need to be able to tell, for example, a Gigabit PHY to only advertise 100M FDX or configurations like "100M FDX and anything below that" even if the link partner also has GBit capability.
- The PHY driver implementation to be used in a certain target is specified via the PHY's "compatible" property in the device tree. I personally need auto-detection. We're a large scale device manufacturer, and our PCB manufacturer was given a list of pin-compatible, tested PHYs that is updated whenever a part on that list reaches EOL status. For each production run, they'll use whatever PHY is currently in their parts bin. Therefore, two devices that look identical on the outside may have anything between subtle differences within a PHY register all the way to a different register layout.
- There's also one particular corner case this implementation doesn't consider - but neither does mine yet: Each GEM, in both the Zynq and the UltraScale, has its own MDIO bus, which is accessed via the respective GEM's register space. Yet, we have a device which uses GEM0 & GEM1, each connected to its own physical Ethernet port. Yet, there just weren't enough free I/O pins in order to expose the MDIO interface of GEM1. Therefore, both PHYs are attached to the MDIO bus of GEM0. Which means that not only does each GEM instance need to know the address of its respective PHY on the MDIO bus, but also the address of the base device via which it can be accessed. Considering that the device driver instances of both GEMs might want to talk to their respective PHYs simultaneously, this scenario will eventually require a mutex concept for shared MDIO access.

Reply:
Another feature the suggested PHY framework is currently missing: LED configuration.